src/controller/proxy/manifestcache.go

```diff
@@ -0,0 +1,258 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/manifest/manifestlist"
+	"github.com/docker/distribution/manifest/schema2"
+	"github.com/goharbor/harbor/src/lib"
+	libCache "github.com/goharbor/harbor/src/lib/cache"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"strings"
+	"time"
+)
+
+const defaultHandler = "default_handler"
+
+// NewCacheHandlerRegistry ...
+func NewCacheHandlerRegistry(local localInterface) map[string]ManifestCacheHandler {
+	registry := map[string]ManifestCacheHandler{
+		manifestlist.MediaTypeManifestList: &ManifestListCache{local: local, cache: libCache.Default(), listProcessor: ListProcessor{local, libCache.Default()}},
+		v1.MediaTypeImageIndex:             &ImageIndexCache{local, ListProcessor{local, libCache.Default()}},
+		schema2.MediaTypeManifest:          &ManifestCache{local},
+		"default":                          &ManifestCache{local},
+	}
+	return registry
+}
+
+// ManifestCacheHandler define how to cache manifest content
+type ManifestCacheHandler interface {
+	// CacheContent - cache the content of the manifest
+	CacheContent(ctx context.Context, remoteRepo string, man distribution.Manifest, art lib.ArtifactInfo, contType string, r RemoteInterface)
+	// Push - push the manifest to local registry
+	Push(ctx context.Context, repo, reference string, man distribution.Manifest) error
+}
+
+// ManifestListCache handle Manifest list type
+type ManifestListCache struct {
+	cache         libCache.Cache
+	local         localInterface
+	listProcessor ListProcessor
+}
+
+// CacheContent ...
+func (m *ManifestListCache) CacheContent(ctx context.Context, remoteRepo string, man distribution.Manifest, art lib.ArtifactInfo, contType string, r RemoteInterface) {
+	_, payload, err := man.Payload()
+	if err != nil {
+		log.Errorf("failed to get payload, error %v", err)
+		return
+	}
+	key := getManifestListKey(art.Repository, art.Digest)
+	log.Debugf("ache manifest list with key=cache:%v", key)
+	err = m.cache.Save(key, payload, manifestListCacheInterval)
+	if err != nil {
+		log.Errorf("failed to cache payload, error %v", err)
+	}
+	err = m.Push(ctx, art.Repository, getReference(art), man)
+	if err != nil {
+		log.Errorf("error when push manifest list to local :%v", err)
+	}
+	return
+}
+
+func (m *ManifestListCache) updateManifestList(ctx context.Context, repo string, manifest distribution.Manifest) (distribution.Manifest, error) {
```

Comment:
This is dup with `ListProcessor.UpdateManifestList`

Reply:
removed